runtime/domain.c

```diff
@@ -1349,7 +1345,7 @@ CAMLprim value caml_domain_spawn(value callback, value term_sync)
   /* When domain 0 first spawns a domain, the backup thread is not active, we
      ensure it is started here. */
   domain_self->tid = pthread_self();
-  if (atomic_load_acquire(&domain_self->backup_thread_running) == 0)
+  if (!backup_thread_running(domain_self))
     install_backup_thread(domain_self);
```

Comment:
Note: the logic here feels a bit fragile to me, because in general a backup thread can be running that comes from an older domain (an older instance of the same domain slot) that is in the process of terminating. This should not happen in the case of domain 0 that we are thinking about here, but I don't see any robust check that this is in fact the first spawn from domain 0, so from a distance it looks like we may fail to register a backup thread for some newly spawned domain if a previous domain is not done with its own backup thread in the same slot.